crates/core/tedge_api/src/plugin.rs

```diff
@@ -0,0 +1,619 @@
+/// passed to the plugin is sound (what "sound" means in this context is to be decided by the
+/// plugin author, i.e. the author of the implementation of this trait).
+///
+/// The plugin author must also name all message types the plugin which is about to be instantiated
+/// can receive (see `PluginBuilder::kind_message_types`).
+#[async_trait]
+pub trait PluginBuilder<PD: PluginDirectory>: Sync + Send + 'static {
+    /// The name for the kind of plugins this creates, this should be unique and will prevent startup otherwise
+    ///
+    /// The "kind name" of a plugin is used by the configuration to name what plugin is to be
+    /// instantiated. For example, if the configuration asks thin-edge to instantiate a plugin
+    /// of kind "foo", but only a plugin implementation of kind "bar" is compiled into thin-edge,
+    /// the software is able to report misconfiguration on startup.
+    fn kind_name() -> &'static str
+    where
+        Self: Sized;
+
+    /// A list of message types the plugin this builder creates supports
+    ///
+    /// This function must return a `HandleTypes` object which represents the types of messages
+    /// that a plugin is able to handle.
+    ///
+    /// To create an instance of this type, you must use the [`HandleTypes::get_handlers_for`] method.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// # use tedge_api::{Plugin, plugin::BuiltPlugin, PluginError, PluginDirectory, PluginBuilder, PluginConfiguration};
+    ///
+    /// #[derive(Debug)]
+    /// struct MyMessage;
+    /// impl tedge_api::plugin::Message for MyMessage {
+    ///     type Reply = tedge_api::message::NoReply;
+    /// }
+    ///
+    /// struct MyPluginBuilder;
+    /// struct MyPlugin; // + some impl Plugin for MyPlugin
+    /// # #[async_trait::async_trait]
+    /// # impl Plugin for MyPlugin {
+    /// #     async fn setup(&mut self) -> Result<(), PluginError> {
+    /// #         unimplemented!()
+    /// #     }
+    /// #     async fn shutdown(&mut self) -> Result<(), PluginError> {
+    /// #         unimplemented!()
+    /// #     }
+    /// # }
+    ///
+    /// #[async_trait::async_trait]
+    /// impl tedge_api::plugin::Handle<MyMessage> for MyPlugin {
+    ///     async fn handle_message(
+    ///         &self,
+    ///         message: MyMessage,
+    ///         sender: tedge_api::address::ReplySender<tedge_api::message::NoReply>
+    ///     ) -> Result<(), tedge_api::error::PluginError> {
+    ///         // ... Do something with it
+    ///#         Ok(())
+    ///     }
+    /// }
+    ///
+    /// #[async_trait::async_trait]
+    /// impl<PD: PluginDirectory> PluginBuilder<PD> for MyPluginBuilder {
+    ///     fn kind_message_types() -> tedge_api::plugin::HandleTypes
+    ///     where
+    ///         Self: Sized,
+    ///     {
+    ///         tedge_api::plugin::HandleTypes::declare_handlers_for::<(MyMessage,), MyPlugin>()
```

Comment:
This example might improved using more than one message type. 


```
    ///         tedge_api::plugin::HandleTypes::declare_handlers_for::<(MyMessage1,MyMessage2,MyMessage3), MyPlugin>()
```

This would explain better:
* why is `declare_handlers_for` required
* that the plugin has to `impl Handler<>` for all these message types.